File: src/aoc/day_6.rs
use std::collections::HashMap;
use std::fs;
use std::str::FromStr;
use itertools::Itertools;
use crate::aoc::MissingFileErr;

pub fn day_6() {
  let file_name = "src/resources/day_6_input.txt";
  println!("Day 6 answer: {}", day_6_part_1_answer(file_name));
  println!("Day 6 answer: {}", day_6_part_2_answer(file_name));
}


#[derive(Hash, Eq, PartialEq, Clone)]
struct MemoKey(LanternFish, Days);
type Days = u64;


fn get_num_fishies_fish_makes(
  fish: LanternFish,
  days: u64,
  memo: &mut HashMap<MemoKey, u64>
) -> u64 {

  let key = MemoKey(fish.clone(), days);
  if let Some(num_fishies) = memo.get(&key) {
    return num_fishies.clone()
  }

  if days == 0 {
    memo.insert(key, 1);
    return 1

  }

  return if fish.timer == 0 {
    let amount =
      get_num_fishies_fish_makes(LanternFish::new_baby(), days-1, memo) +
      get_num_fishies_fish_makes(fish.reset(), days-1, memo);

    memo.insert(key, amount.clone());
    amount
  } else {
    get_num_fishies_fish_makes(fish.subtract_day(), days-1, memo)
  };
}

fn get_total_number_fishies_created(mut fishies: Vec<LanternFish>, days: u64) -> u64 {
  let mut num_can_create: HashMap<MemoKey, u64> = HashMap::new();

  fishies
    .iter()
    .fold(0u64, |amount, fish| {
      amount + get_num_fishies_fish_makes(fish.clone(), days.clone(), &mut num_can_create)
  })
}


#[derive(Clone, Hash, Eq, PartialEq)]
struct LanternFish {
  timer: u64
}

impl LanternFish {
  fn new_baby() -> LanternFish {
    LanternFish { timer: 8 }
  }

  fn reset(&self) -> LanternFish {
    LanternFish { timer: 6 }
  }

  fn subtract_day(&self) -> LanternFish {
    LanternFish { timer: self.timer - 1 }
  }

  fn new(timer: u64) -> LanternFish {
    LanternFish { timer }
  }
}

fn parse_input_file(file_name: &str) -> Result<Vec<LanternFish>, MissingFileErr> {
  let input =
    fs::read_to_string(file_name)
      .map_err(|_| MissingFileErr(file_name.to_string()))?;

  let fish = input
    .split(',')
    .map(u64::from_str)
    .filter_map(Result::ok)
    .map(LanternFish::new)
    .collect_vec();

  Ok(fish)
}

pub fn day_6_part_1_answer(file_name: &str) -> u64 {
  let mut fishies = parse_input_file(file_name).unwrap();
  get_total_number_fishies_created(fishies, 80)
}

pub fn day_6_part_2_answer(file_name: &str) -> u64 {
  let mut fishies = parse_input_file(file_name).unwrap();

  get_total_number_fishies_created(fishies, 256)
}

File: kotlin/src/main/kotlin/advent/of/code/Day-6.kt
package advent.of.code

private data class LanternFish(val timer: Int) {
  fun subtractDay() = LanternFish(timer - 1)

  companion object {
    fun baby() = LanternFish(8)
    fun reset() = LanternFish(6)
  }
}

private data class MemoKey(val fish: LanternFish, val daysLeft: Int)

private fun getNumberFishiesMade(
  fish: LanternFish,
  daysLeft: Int,
  stored: MutableMap<MemoKey, Long>
): Long {
  val key = MemoKey(fish, daysLeft)
  stored[key]?.let{ return it }

  val numCreated = if (daysLeft == 0) {
    1L
  } else if (fish.timer == 0) {
    getNumberFishiesMade(LanternFish.reset(), daysLeft-1, stored) + getNumberFishiesMade(LanternFish.baby(), daysLeft - 1, stored)
  } else {
    getNumberFishiesMade(fish.subtractDay(), daysLeft-1, stored)
  }

  return numCreated.also{ stored[key] = it }
}

fun getTotalFishiesMade(input: String, daysLeft: Int): Long {
  val fishies = input.split(",").map{ LanternFish(it.toInt()) }

  val storedVals = mutableMapOf<MemoKey, Long>()
  return fishies.fold(0) { numFishies, fish ->
    numFishies + getNumberFishiesMade(fish, daysLeft, storedVals)
  }
}

object Day6 {
  val fileName = "day-6-input.txt"

  fun part1Answer(input: String) = getTotalFishiesMade(input, 80)

  fun part2Answer(input: String) = getTotalFishiesMade(input, 256)

  fun answers() {
    readFileToString(fileName)?.let{
      println("Day 6 Part 1: ${part1Answer(it)}")
      println("Day 6 Part 2: ${part2Answer(it)}")
    }
  }
}